packages/react/src/ReactRenderer.tsx

```diff
@@ -120,14 +148,32 @@ export class ReactRenderer<R = unknown, P extends Record<string, any> = object>
     const props = this.props
     const editor = this.editor as EditorWithContentComponent
 
-    if (isClassComponent(Component) || isForwardRefComponent(Component)) {
-      // @ts-ignore This is a hack to make the ref work
-      props.ref = (ref: R) => {
-        this.ref = ref
+    // Handle ref forwarding with React 18/19 compatibility
+    const isReact19 = isReact19Plus()
+    const isClassComp = isClassComponent(Component)
+    const isForwardRefComp = isForwardRefComponent(Component)
+
+    const elementProps = { ...props }
+
+    if (!elementProps.ref) {
+      if (isReact19) {
+        // React 19: ref is a standard prop for all components
+        // @ts-ignore - Setting ref prop for React 19 compatibility
+        elementProps.ref = (ref: R) => {
+          this.ref = ref
+        }
+      } else if (isClassComp || isForwardRefComp) {
+        // React 18 and prior: only set ref for class components and forwardRef components
+        // @ts-ignore - Setting ref prop for React 18 class/forwardRef components
+        elementProps.ref = (ref: R) => {
+          this.ref = ref
+        }
       }
+      // For function components in React 18, we can't use ref - the component won't receive it
+      // This is a limitation we have to accept for React 18 function components without forwardRef
     }
 
-    this.reactElement = <Component {...props} />
+    this.reactElement = <Component {...elementProps} />
```

Comment:
Something still seems off here in React 18. I just bumped TipTap from `beta.5` to `beta.6` on a project, and my custom TipTap React nodes are logging the following ref warning regardless of whether I wrap my `TestComponent` render function in a `forwardRef` or not.

Warning:
```
Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()? Error Component Stack
    at ReactNodeView (<anonymous>)
    ...
```

`reactVersion` is logging as `18.3.1`

One of my components' render functions is
```
  addNodeView() {
    return ReactNodeViewRenderer(TestComponent)
  },
  ```

Reply:
Thanks, will take a look into that

Reply:
Yeah, I'm getting the same thing with upgrading from 2.12 to 2.14

I'm on React 18.